### PR TITLE
Initial refactor for list command

### PR DESCRIPTION
**Pull request by baduker**
**Refactor `download.go` in Preparation for [Add list command](https://github.com/exercism/cli/pull/1113)**

**Purpose of this PR:**

This PR primarily aims to simplify the `download` command, paving the way for the seamless integration of the `list` command. This will eventually allow users to list tracks and related exercises (refer to the linked PR for details).

**Key Modifications (ordered for clarity)**:

**1. `download.go`:**
   - Introduced a shared `*api.Client` across all methods and functions for uniformity.
   - Decomposed `runDownload` and `newDownload` functions to enhance readability.
   - Eliminated superfluous `usrCfg` verifications.
   - Extracted `usrCfg` items from `solutionDownload` structs, decoupling configuration from download operations.
   - Modularized and delegated the setup and validation of the `download` command flags to distinct methods.
   - Unified the creation process of `solutionURL`.
   - Refactored method names, replacing vague identifiers like `url()`, `files()`.
   - Reorganized the structure: placed all structs at the beginning, and logically grouped related methods/functions.
   - Augmented all functions, methods, and structs with descriptive comments.

**2. `client.go`:**
   - Introduced `MakeRequest` to merge the functionalities of `client.NewRequest` and `client.Do`, enhancing the DRY principle.

**3. `configure.go`:**
   - Implemented `LoadUserConfig()`, which encapsulates the logic for fetching the user config governed by Viper.

---

**Compatibility & Testing:**

To the best of my knowledge, this PR introduces no breaking changes. All previously established functionalities should persist undisturbed. Both the `go test ./...` command on this branch and the test suite in `download_test.go` have been executed without any discrepancies.

### CODE REVIEW

**Comment by iHiD** [comment]
@baduker Thanks!
@nywilken Would you be happy to review this pls?

**Comment by nywilken** [comment]
> @baduker Thanks! @nywilken Would you be happy to review this pls?

@iHiD I can take a look this week for sure.

**Comment by nywilken** [comment]
Hi @baduker thanks for opening up this change. I've started reviewing but won't be able to complete today. I'll pick it up on Monday. 

So far I see what you did with the MakeRequest function but I find the function signature a little confusing as it doesn't indicate that providing false to isFullUrl will automatically append the API URL or that the MakeRequest will always make a Get request. 

I'll give it a more complete review on Monday. In the mean, time we've updated the minimum Go version to 1.20 could you rebase your changes on to the latest main to get the test passing again. Thanks.

**Comment by baduker** [comment]
Hi @nywilken! Thanks for looking at this. Fair point about the `MakeRequest` function. I'll see what I can do to improve the signature and make the `isFullUrl` more explicit. As for the upgraded to `Go 1.20`, I've already rebased and `force-pushed`, however, I don't get the checks executed because the workflow is awaiting approval. Any idea why that is?

Looking forward to your full review and the discussion!

**Comment by baduker** [comment]
Hi @nywilken! Thanks for looking at this. Fair point about the `MakeRequest` function. I'll see what I can do to improve the signature and make the `isFullUrl` more explicit. As for the upgraded to `Go 1.20`, I've already rebased and `force-pushed`, however, I don't get the checks executed because the workflow is awaiting approval. Any idea why is that?

**Comment by nywilken** [comment]
> I've already rebased and force-pushed, however, I don't get the checks executed because the workflow is awaiting approval. Any idea why that is?
Looking forward to your full review and the discussion!

@baduker apologies for the delay. I wasn't able to make time on Monday. I've allocated time for my Friday. Apologies again. 

As for the the required checks, I believe their is a repo setting for PRs that I don't have access to change. Maybe this is something that @ErikSchierboom can assist with when he has a moment.

**Comment by baduker** [comment]
Hi @nywilken! Thanks for keeping me posted and don't worry too much about it. We all do open software in our spare time and sometimes it's just not enough of it.

Anyways, drop me a line when you're done reviewing and I'll get back to you as soon as I can.

File: cmd/configure.go
Comment (nywilken) [inline cmd/configure.go:40]: Comparing the logic from the configure and download commands it looks like the download command is now modifying the global viperConfig variable when it doesn't need to.

 Consolidating the loading of viper into a shared function makes sense but it seems like it has to happen outside of the configure command and be a bit more explicit on its intent. 

Looking at all the commands I see the general flow for the initializing the user config is as follows:
- create NewConfig
- Initialize instance of Viper
- Read viper config
- validate user config (error if invalid)
- done

I think the shared function should do just that but either be renamed to indicate it is doing that or be rewritten just to handle the viper loading bits. For example I can see this function moved into the config package with a new function signature

```go
func LoadUserConfigFile() (config.Config, error) {
	cfg := config.NewConfig()
	v := viper.New()
	v.AddConfigPath(cfg.Dir)
	v.SetConfigName("user")
	v.SetConfigType("json")
	// Ignore error. If the file doesn't exist, that is fine.
	_ = v.ReadInConfig()
	cfg.UserViperConfig = v

	if err := validateUserConfig(cfg.UserViperConfig); err != nil {
		return nil, err
	}
	return cfg, nil
}
```
Or Just breakout the viper config parts
```go
func LoadViperConfig(dir, name string) (*viper.Viper, error) {
        v := viper.New()
	v.AddConfigPath(dir)
	v.SetConfigName("user")
	v.SetConfigType("json")
	//Maybe capture and return the error 
	_ = v.ReadInConfig()

	return v, nil
}

// Which would change its usage to something like 
cfg := NewConfig()
cfg.UserViperConfig, err := LoadViperConfig(cfg.Dir, "user")
if err != nil {
  /// handle
}
if err := validateUserConfig(cfg.UserViperConfig); err != nil {
   //handle error as needed for the command
}

```

**Review by nywilken — CHANGES_REQUESTED** [review]
@baduker thanks again for the change and for your understanding with my time to review.

 I get the approach but I think we may need to break up the change a bit more to help provide small actionable suggestions. I find a few of the function rewrites to duplicate the data being passed along like client and userCfg in the download command. 

 But to better assist with the review could you share what parts of the code you think could benefit from a readability rewrite and start there. 

For the shared code I can see you wanting to reuse for the list command so that makes sense and maybe starting there will help clean things up a bit to start. 

Also I just want to say that my feedback is not a hard stance and we can definitely iterate on things. 

I look forward to hearing your thoughts.

**Comment by baduker** [comment]
Hey @nywilken! Thanks for the review and feedback. I haven't really gotten around to it yet but give me a couple days and I'll get back to you some changes and points for discussion.

**Comment by baduker** [comment]
Hello @nywilken! I've been unwell recently, but I'm back now, full of energy and ready to go. I need a little time to refresh my memory on the changes. I'll get back to you with an update in a few days. Thank you for your patience!

**Comment by nywilken** [comment]
> Hello @nywilken! I've been unwell recently, but I'm back now, full of energy and ready to go. I need a little time to refresh my memory on the changes. I'll get back to you with an update in a few days. Thank you for your patience!

@baduker glad to hear you are well and on the mend ❤️‍🩹. Please take care and ping whenever you’re ready.

Things on the PR front seem to be on hold for a bit as the team’s priorities are focused on the platform. I will provide any guidance I can but I cant merge or fix the required checks issue for Go 1.15.

File: cmd/download.go
Comment (baduker) [reply cmd/download.go]: To be honest, I don't remember why I pulled out those variables. You're right, they're used once, and there's no need to define them there at this point.

File: cmd/download.go
Comment (baduker) [reply cmd/download.go:91]: Well, this is not a bad idea, but if I move `LoadUserConfig()` into `runDownload()`, then I'd have to change the `download_test.go`. Should I go for that change, too?

File: cmd/configure.go
Comment (baduker) [reply cmd/configure.go]: I took your first suggestion, expanding it a bit.

File: api/client.go
Comment (baduker) [reply api/client.go]: I improved the function, taking into consideration your input.

**Comment by baduker** [comment]
Hello @nywilken! I trust this message finds you in good health. I've implemented most of your suggestions for the PR and have updated my changes. However, there's one thread that remains unresolved, as I require further input from you. I eagerly await your response!

**Comment by baduker** [comment]
Hello @nywilken! I hope you're doing well. I wanted to check in with you regarding the status of this pull request. I'm eager to complete this if you still see the need for the `list` command. Please let me know! Take care!

**Comment by baduker** [comment]
@nywilken is this thread still alive? @ErikSchierboom this has stalled a bit. Should I drop the changes or are you still up for the enhancement?

**Comment by ErikSchierboom** [comment]
In a previous comment, @nywilken said:

> I get the approach but I think we may need to break up the change a bit more to help provide small actionable suggestions

I agree with this. This PR is quite a big refactoring, which makes it hard to review. I'm not opposed to some refactoring/cleanup, but I'd like them to be as minimal as possible.

I'm not well versed into the CLI code base. Maybe @ekingery could also take a look?

**Comment by ErikSchierboom** [comment]
@baduker I'm going over all the PRs and whilst browsing this PR, I think my main concern is that it is doing so many different things. Whilst there are definitely things in this PR worth exploring, I'm gonna close this PR. 

If you'd still like to work on this codebase, please open an issue on the forum first: https://forum.exercism.org/c/exercism/4 We can then discuss what would be worthwhile to work on.

**Pull request closed without merging**
